### PR TITLE
Feature/ATG-63 Create canvas, class Figure, styles to board, logic game (move down)

### PR DESCRIPTION
**Pull request by IraCherginyshka**
Implamenting game logic.
Create class Figure
Create canvas in game-component
Move down figure

### CODE REVIEW

File: src/app/components/game-board/game-board.component.ts [hunk 1]
```diff
@@ -1,10 +1,67 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, ViewChild, ElementRef, OnInit } from '@angular/core';
+import Figure from '../../models/Figure';
+import Block from '../../models/Block';
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: No such property in this branch

 ERROR in src/app/components/game-board/game-board.component.ts(3,19): error TS2307: Cannot find module '../../models/Block'.

File: src/app/components/game-board/game-board.component.html [hunk 2]
```diff
@@ -1 +1,3 @@
-<p>game-board works!</p>
+<div>
+  <canvas #canvas width="325" height="500" class="board-canvas"></canvas>
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.html]: width="325" height="500"  -- "Magic numbers"

File: src/app/components/game-board/game-board.component.ts [hunk 3]
```diff
@@ -1,10 +1,67 @@
...
 @Component({
   selector: 'atg-game-board',
   templateUrl: './game-board.component.html',
   styleUrls: ['./game-board.component.scss'],
 })
 export class GameBoardComponent implements OnInit {
-  ngOnInit() {}
+  @ViewChild('canvas', { static: true })
+  private canvas: ElementRef<HTMLCanvasElement>;
+
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: please do not add en empty line between class members with same access modifier

File: src/app/components/game-board/game-board.component.ts [hunk 4]
```diff
@@ -1,10 +1,67 @@
...
+  private boardMatrix: any[];
+
+  ngOnInit(): void {
+    this.ctx = this.canvas.nativeElement.getContext('2d');
+    this.boardMatrix = this.makeBoardEmptyMatrix(QUANTITY_BLOCKS_WIDTH, QUANTITY_BLOCKS_HEIGHT);
+    this.play();
+  }
+
+  makeBoardEmptyMatrix(width, height) {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: please add types to parameters

Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: Can be private

File: src/app/components/game-board/game-board.component.ts [hunk 5]
```diff
@@ -1,10 +1,67 @@
...
+    let lineNumber = height;
+    const matrix = [];
+    while (lineNumber !== 0) {
+      matrix.push(new Array(width).fill(0));
+      lineNumber--;
+    }
+    return matrix;
+  }
+
+  drawBoard(matrix) {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: can be private

File: src/app/components/game-board/game-board.component.ts [hunk 6]
```diff
@@ -1,10 +1,67 @@
...
+    this.boardMatrix = this.makeBoardEmptyMatrix(QUANTITY_BLOCKS_WIDTH, QUANTITY_BLOCKS_HEIGHT);
+    this.play();
+  }
+
+  makeBoardEmptyMatrix(width, height) {
+    let lineNumber = height;
+    const matrix = [];
+    while (lineNumber !== 0) {
+      matrix.push(new Array(width).fill(0));
+      lineNumber--;
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: get no-pluplus error for es lint for this row, please disable this rule

File: src/app/components/game-board/game-board.component.ts (hunk 4 above)
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: add return type

File: src/app/components/game-board/game-board.component.ts (hunk 5 above)
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: add parameter type and return value

File: src/app/components/game-board/game-board.component.ts [hunk 7]
```diff
@@ -1,10 +1,67 @@
...
+
+  drawBoard(matrix) {
+    matrix.forEach((line, indexY) => {
+      line.forEach((item, indexX) => {
+        if (!item) {
+          const block = new Block(this.ctx, 'darkblue');
+          block.fillBoardBlock(indexX, indexY);
+        } else {
+          const block = new Block(this.ctx, 'red');
+          block.fillBoardBlock(indexX, indexY);
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: duplicate of line 44, extract out of if else statements

File: src/app/components/game-board/game-board.component.ts [hunk 8]
```diff
@@ -1,10 +1,67 @@
...
+        }
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: The only difference of this block is color parameter - can be refactored into 
`const color =  item ? 'color1' : 'color2';
new Block(this.ctx, color).fillBoardBlock(indexX, indexY);

`

File: src/app/components/game-board/game-board.component.ts [hunk 9]
```diff
@@ -1,10 +1,67 @@
...
+      });
+    });
+  }
+
+  play() {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: add return type

File: src/app/components/game-board/game-board.component.ts [hunk 10]
```diff
@@ -1,10 +1,67 @@
...
+    let height = 0;
+    const newFigure = new Figure();
+    const { figureMatrix } = newFigure;
+
+    const interval = setInterval(() => {
+      this.drawBoard(newFigure.showFigure(height, this.boardMatrix));
+      if (height + figureMatrix.length === 20) {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: 20 is magic number

File: src/app/components/game-board/game-board.component.ts [hunk 11]
```diff
@@ -1,10 +1,67 @@
...
+          block.fillBoardBlock(indexX, indexY);
+        }
+      });
+    });
+  }
+
+  play() {
+    let height = 0;
+    const newFigure = new Figure();
+    const { figureMatrix } = newFigure;
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: we only using length of figureMatrix, so i would write 

`const itemHeight = newFigure.figureMatrix.length `; 

and use itemHeight later

File: src/app/models/Figure.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { CENTRAL_ITEM } from '../constants/board-component-constants';
+
+export default class Figure {
+  public figureMatrix = [[1, 1], [1, 1]];
+
+  showFigure(height: number, boardMatrix: any[]) {
```
Comment (oleksandrstarov) [inline src/app/models/Figure.ts]: please add return type

File: src/app/models/Figure.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    const matrixWithFigure = boardMatrix.slice();
+    let figureLine = 0;
+    for (let i = height; i < this.figureMatrix.length + height; i++) {
+      const lineWithFigure = boardMatrix[i].slice();
+      lineWithFigure.splice(
+        CENTRAL_ITEM,
+        this.figureMatrix[figureLine].length,
+        ...this.figureMatrix[figureLine],
```
Comment (oleksandrstarov) [inline src/app/models/Figure.ts]: Please check what is returned by this line

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/app/components/game-board/game-board.component.ts (hunk 1 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: I fixed it

File: src/app/components/game-board/game-board.component.ts (hunk 3 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: EsLint
I'll delete this rule

File: src/app/components/game-board/game-board.component.html (hunk 2 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.html]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 4 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 9 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 4 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 10 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 11 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/models/Figure.ts (hunk 12 above)
Comment (IraCherginyshka) [reply src/app/models/Figure.ts]: Done

File: src/app/models/Figure.ts (hunk 13 above)
Comment (IraCherginyshka) [reply src/app/models/Figure.ts]: This line returns a list of items, if I delete spread operator, I'll add array to the selected line.
Is it wrong?

File: src/app/components/game-board/game-board.component.ts [hunk 14]
```diff
@@ -1,10 +1,56 @@
...
 @Component({
   selector: 'atg-game-board',
   templateUrl: './game-board.component.html',
   styleUrls: ['./game-board.component.scss'],
 })
 export class GameBoardComponent implements OnInit {
-  ngOnInit() {}
+  @ViewChild('canvas', { static: true }) private canvas: ElementRef<HTMLCanvasElement>;
+  private ctx: CanvasRenderingContext2D;
+  private boardMatrix: any[];
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: This variable value clearly generated by makeBoardEmptyMatrix, I cant see any reason not to assign FiguresColors[][]  type to it

File: src/app/components/game-board/game-board.component.ts [hunk 15]
```diff
@@ -1,10 +1,56 @@
...
+
+  ngOnInit(): void {
+    this.canvas.nativeElement.width = CANVAS_WIDTH;
+    this.canvas.nativeElement.height = CANVAS_HEIGHT;
+    this.ctx = this.canvas.nativeElement.getContext('2d');
+    this.boardMatrix = this.makeBoardEmptyMatrix(QUANTITY_BLOCKS_WIDTH, QUANTITY_BLOCKS_HEIGHT);
+    this.play();
+  }
+
+  makeBoardEmptyMatrix(width: number, height: number): FiguresColors[][] {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: Can be private

File: src/app/components/game-board/game-board.component.ts [hunk 16]
```diff
@@ -1,10 +1,56 @@
...
+    return new Array(height).fill(new Array(width).fill(FiguresColors.DEFAULT));
+  }
+
+  play(): void {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: Can be private

File: src/app/components/game-board/game-board.component.ts [hunk 17]
```diff
@@ -1,10 +1,56 @@
...
+
+  private drawBoard(matrix: FiguresColors[][]): void {
+    matrix.forEach((line, indexY) => {
+      line.forEach((item, indexX) => {
+        if (item === FiguresColors.DEFAULT) {
+          const block = new BlockModel(this.ctx, FiguresColors.DEFAULT);
+          block.fillBoardBlock(indexX, indexY);
+        } else {
+          const block = new BlockModel(this.ctx, item);
+          block.fillBoardBlock(indexX, indexY);
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: both are same. Looks strange. can be 
`line.forEach((item, indexX) => new BlockModel(this.ctx, item).fillBoardBlock(indexX, indexY)`

File: src/app/models/figure.model.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    const matrixWithFigure = [...boardMatrix];
+    let figureLine = 0;
+    for (let i = height; i < this.figureMatrix.length + height; i++) {
+      const lineWithFigure = [...boardMatrix[i]];
+      lineWithFigure.splice(
+        CENTRAL_ITEM,
+        this.figureMatrix[figureLine].length,
+        ...this.figureMatrix[figureLine],
+      );
+      matrixWithFigure[i] = lineWithFigure.slice();
```
Comment (oleksandrstarov) [inline src/app/models/figure.model.ts]: why not to keep using `= [..lineWithFigere]` ?

File: src/app/models/figure.model.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+      figureLine += 1;
+    }
+    return matrixWithFigure;
+  }
```
Comment (oleksandrstarov) [inline src/app/models/figure.model.ts]: can be changed into, what do you think?
`showFigure(height: number, boardMatrix: FiguresColors[][]): FiguresColors[][] {
    const matrixWithFigure = [...boardMatrix];

    this.figureMatrix.forEach((line, index) => {
      const targetLine = [...boardMatrix[index + height]];
      targetLine.splice(CENTRAL_ITEM, line.length, ...line);
      matrixWithFigure[index + height] = [...targetLine];
    });
    return matrixWithFigure;
  }
`

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/app/components/game-board/game-board.component.ts (hunk 14 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 16 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 17 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/models/figure.model.ts (hunk 18 above)
Comment (IraCherginyshka) [reply src/app/models/figure.model.ts]: True

File: src/app/models/figure.model.ts (hunk 19 above)
Comment (IraCherginyshka) [reply src/app/models/figure.model.ts]: Красиво
I changed

File: src/app/components/game-board/game-board.component.ts (hunk 15 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

**Review by oleksandrstarov — APPROVED** [review]

**Pull request merged**